gpspipe.1
'\" t
.\"     Title: gpspipe
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 03 Aug 2005
.\"    Manual: GPSD Documentation
.\"    Source: The GPSD Project
.\"  Language: English
.\"
.TH "GPSPIPE" "1" "03 Aug 2005" "The GPSD Project" "GPSD Documentation"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gpspipe \- tool to connect to gpsd and retrieve sentences
.SH "SYNOPSIS"
.HP \w'\fBgpspipe\fR\ 'u
\fBgpspipe\fR [\-h] [\-d] [\-l] [\-o\ \fIfilename\fR] [\-n\ \fIcount\fR] [\-r] [\-R] [\-s\ \fIserial\-device\fR] [\-t] [\-T\ \fItimestamp\-format\fR] [\-u] [\-p] [\-w] [\-S] [\-2] [\-v] [\-D\ \fIdebug\-level\fR] [\fIserver\fR [\fI:port\fR [\fI:device\fR]]]
.SH "DESCRIPTION"
.PP
gpspipe
is a tool to connect to
gpsd
and output the received sentences to stdout\&. This makes the program useful as a pipe from
gpsd
to another program or file\&.
.PP
gpspipe
does not require root privileges, and can be run concurrently with other tools connecting to the local
gpsd
without causing problems\&.
.PP
The output will consist of one or both of the raw NMEA or native
gpsd
sentences\&. Each line can be optionally time stamped\&. There is also an option to exit gracefully after a given count of packets\&.
.PP
Optionally a server, TCP/IP port number and remote device can be given\&. If omitted,
gpspipe
connects to localhost on the default port (2947) and watches all devices opened by
gpsd\&.
.PP
gpspipe
may be run as a daemon, but requires the \-o flag for writing the output to a file\&.
.SH "OPTIONS"
.PP
\-h makes
gpspipe
print a usage message and exit\&.
.PP
\-d causes
gpspipe
to run as a daemon\&.
.PP
\-l causes
gpspipe
to sleep for ten seconds before attempting to connect to gpsd\&. This is very useful when running as a daemon, giving gpsd time to start before attempting a connection\&.
.PP
\-r causes raw NMEA sentences to be output\&.
.PP
\-R causes super\-raw (gps binary) data to be output\&. This overrides NMEA and gpsd output modes\&.
.PP
\-s option causes the collected data to be written to the specified serial device with settings 4800 8N1\&. Thus
gpspipe
can be used with \-s and \-r options to emulate a serial port hardwired to a GPS that
gpsd
is managing\&.
.PP
\-o option causes the collected data to be written to the specified file\&. Use of this option is mandatory if
gpspipe
is run as a daemon\&.
.PP
\-w causes native
gpsdsentences to be output\&.
.PP
\-S sets the scaled flag\&.
.PP
\-2 sets the split24 flag on AIS reports\&. Note: this option is experimental and may be changed or removed in a future release\&.
.PP
\-t adds a timestamp to each sentence output\&.
.PP
\-T sets the format of the timestamp\&. See
\fBstrftime\fR(3)
for the available placeholders\&. Setting this option implies \-t\&. Default setting is "%F %T"
.PP
\-u usec resolution time stamp, implies \-t\&. Use \-uu to output sec\&.usec\&.
.PP
\-p enables dumping of profiling information in JSON\&.
.PP
\-P enables dumping of PPS drift JSON in NMEA and raw modes\&.
.PP
\-n [count] causes [count] sentences to be output\&.
gpspipe
will then exit gracefully\&.
.PP
\-v causes
gpspipe
to show a spinning activity indicator on stderr\&. This is useful if stdout is redirected into a file or a pipe\&. By default the spinner is advanced with every messages written; specifying \-v more than once will double the number of messages required to rotate the spinner\&.
.PP
\-V prints the version, then exits\&.
.PP
At least one of \-R, \-r or \-w must be specified\&.
.SH "EXAMPLE"
.PP
When
gpsd is running
\fBgpspipe \-r \-n 100\fR
will send one hundred raw NMEA sentences to standard output, then exit\&.
.SH "SEE ALSO"
.PP
\fBgpsd\fR(8),
\fBgps\fR(1),
\fBlibgps\fR(3),
\fBlibgpsmm\fR(3),
\fBgpsprof\fR(1),
\fBgpsfake\fR(1),
\fBgpsctl\fR(1),
\fBgpscat\fR(1)\&.
\fBgpsmon\fR(1)\&.
.SH "AUTHOR"
.PP
Gary E\&. Miller
<gem@rellim\&.com>\&.


debian/changelog
gpsd (3.17-3) unstable; urgency=medium

  * [2638b3e] Avoid duplicate ppscheck installation.
    Thanks to Andreas Beckmann (Closes: #879951)

 -- Bernd Zeimetz <bzed@debian.org>  Fri, 27 Oct 2017 17:26:55 +0200

gpsd (3.17-2) unstable; urgency=medium

  * Uploading to unstable.
  * [a13df427] Fixing last changelog entry.
    Some lines went missing, oops.

 -- Bernd Zeimetz <bzed@debian.org>  Thu, 26 Oct 2017 23:16:41 +0200

gpsd (3.17-1) experimental; urgency=medium

  * [d21fbf15] Updated version 3.17 from 'upstream/3.17'
    with Debian dir 3099dc33257e902647917f2ad7e67eec078b0b42
  * [913ea95c] Fix SConstruct.
  * [d9d7b300] Python dbg _d.so files are not built anymore.
  * [f0ea0948] Install ppscheck + manpage
  * [7df3aa0a] Updating symbols files

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 21 Oct 2017 16:12:28 +0200

gpsd (3.16-4) unstable; urgency=medium

  * [2405f456] Fix debug symbol handling.
    Debug symbols went missing when upstream switched the strip
    option name.
    Thanks to Andre Naujoks (Closes: #842792)
  * [74c3cd07] Updating debian/control from debian/control.in
  * [b66410da] hadle debian/control creation a bit better
  * [06a9623f] Use gz compression for now.
  * [44376877] Remove obsolete build dependency on hardening-includes.

 -- Bernd Zeimetz <bzed@debian.org>  Fri, 04 Nov 2016 23:35:14 +0100

gpsd (3.16-3) unstable; urgency=medium

  * [aa43cc65] Various patches were applied upstream.
    Take them from there, removing our version.
  * [1444f940] Build with -std=gnu++98. (Closes: #830993)
  * [505d032a] Remove netbase/systemd-sysv dependency.
    Not needed anymore. (Closes: #768100)
  * [b53473f0] Migrate to github.
  * [cb624dcb] Add travis.yml build stuff.
  * [1c3c3463] Updating gbp.conf to ignore travis.yml for debuild.
  * [d870a5c6] Don't rebuild for the static library.

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 24 Sep 2016 14:34:07 +0200

gpsd (3.16-2) unstable; urgency=medium

  * [5001c11a] Do not start gpsd.service by default.
    We use socket based activation only.
    If one needs a running gpsd.service at boot - without being started by
    gpsd.socket - you'll have to configure systemd to start it.
    Too many people have gpsd installed as dependency and do not need yet
    another running process.

 -- Bernd Zeimetz <bzed@debian.org>  Sun, 15 May 2016 15:35:20 +0200

gpsd (3.16-1) unstable; urgency=medium

  * [c3d16637] Merge tag 'upstream/3.16'
    Upstream version 3.16
  * [29863f7a] Snapshot changelog.
  * [e61a749f] refreshing patches.
  * [ddc25e61] Pick some patches from upstream.
    - Fix test_gpsmm building
    - Fix FTBFS on hurd (Closes: #821804)
  * [67d9aca8] Fix check for clock_gettime.
  * [67b39606] Update git-buildpackage config.

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 30 Apr 2016 20:07:13 +0200

gpsd (3.15-2) unstable; urgency=medium

  * Uploading to unstable

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 31 Oct 2015 20:02:31 +0100

gpsd (3.15-1) experimental; urgency=medium

  * [10e7caf] Install contrib/* as examples.
    Thanks to Sanjeev Gupta (Closes: #781688)
  * [2d150e4] Merge tag 'upstream/3.15'
    Upstream version 3.15
  * [4163f08] snapshot changelog.
  * [1ea8da4] Remove libgpsd.
    Upstream links it statically now and I can't be bothered to maintain
    patches as Gentoo and others do it.
    The result is a slightly larger gpsd-clients package and running
    the binaries in parallel will waste memory.
    Thanks to Paul Wise & Andreas Beckmann (Closes: #779286, #781567)
  * [bb7a556] scons distclean doesn't work anymore.
    use --clean + sconsclean.
  * [31ff968] Refreshing patches
  * [5861e60] Fix broken static linking of test_gpsmm
  * [ef84468] Install new files.
  * [f9e250c] Add patch to fix regression tests for leap second bump
  * [65b8ac6] Updating symbols file.
  * [7e3d4a4] Update libqgpsmm symbols.
  * [785e932] _STRUCT_TIMESPEC is not defined in time.h anymore.
    Only linux/time.h defines _STRUCT_TIMESPEC.
    Check for __timespec_defined additionally.
  * [0b56700] Fix struct member names in timspec.
  * [e723fdd] Use system clock_gettime again.
    Fix some more broken header checks in SConstruct.

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 31 Oct 2015 20:02:07 +0100

gpsd (3.12-1) experimental; urgency=medium

  * [9f0dfbda] Merge tag 'upstream/3.12'
    Upstream version 3.12
  * [0896c861] Refresh/update patches.
  * [9daccea4] Fix test_ debug tool list.
  * [85f4f0bb] Add patch to fix test_gpsmm.
  * [9c511016] Add some new symbols to the symbols file.
  * [30cf1b92] Add dh-python as build dependency.
  * [24c325f9] Bump library sonames in package names.
  * [a810ac32] Update libqgpsmm symbols file
  * [1bf373c6] Fix some lintian errors.

 -- Bernd Zeimetz <bzed@debian.org>  Mon, 23 Feb 2015 22:37:42 +0100

gpsd (3.11-3) unstable; urgency=medium

  [ Andreas Beckmann ]
  * [dd6bbdf1] avoid prompting due to modified lenny conffiles

 -- Bernd Zeimetz <bzed@debian.org>  Sun, 25 Jan 2015 12:15:44 +0100

gpsd (3.11-2) unstable; urgency=medium

  * [a759614c] Avoid prompting due to modified squeeze conffiles.
    Thanks to Andreas Beckmann (Closes: #774872)

 -- Bernd Zeimetz <bzed@debian.org>  Thu, 08 Jan 2015 21:35:30 +0100

gpsd (3.11-1) unstable; urgency=medium

  * [325d1900] Merge tag 'upstream/3.11'
    Upstream version 3.11
  * [8d86b938] Refreshing patches.
  * [877cbcf3] Update symbols file.

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 10 Sep 2014 21:59:28 +0200

gpsd (3.10+dev3~d6b65b48-4) unstable; urgency=medium

  * [8ae93e0]] Two systemd related changes that were forgotton.
    - do not install the old hotplug wrapper
    - don't forget the @ in gpsdctl@.service.
    Thanks to Jeremy Lainé

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 16 Aug 2014 22:12:33 +0200

gpsd (3.10+dev3~d6b65b48-3) unstable; urgency=medium

  * [95fcff8b] Add patch to fix PPS with large offsets.
    Taken from Fedora.
    Thanks to Miroslav Lichvar <mlichvar@redhat.com>

gpspipe.c
/* $Id: gpspipe.c 5053 2009-01-21 11:44:35Z esr $ */
/*
 * gpspipe
 *
 * a simple program to connect to a gpsd daemon and dump the received data
 * to stdout
 *
 * This will dump the raw NMEA from gpsd to stdout
 *      gpspipe -r
 *
 * This will dump the super-raw data (gps binary) from gpsd to stdout
 *      gpspipe -R
 *
 * This will dump the GPSD sentences from gpsd to stdout
 *      gpspipe -w
 *
 * This will dump the GPSD and the NMEA sentences from gpsd to stdout
 *      gpspipe -wr
 *
 * Original code by: Gary E. Miller <gem@rellim.com>.  Cleanup by ESR.
 * All rights given to the gpsd project to release under whatever open source
 * license they use.  A thank you would be nice if you use this code.
 */

#include <sys/types.h>
#include <sys/stat.h>
#ifndef S_SPLINT_S
#include <sys/socket.h>
#include <unistd.h>
#endif /* S_SPLINT_S */
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <fcntl.h>
#include <termios.h>
#include <assert.h>
#include "gpsd_config.h"
#include "gpsd.h"

static int fd_out = 1;		/* output initially goes to standard output */ 
static void spinner(unsigned int, unsigned int);

/* NMEA-0183 standard baud rate */
#define BAUDRATE B4800

/* Serial port variables */
static struct termios oldtio, newtio;
static char serbuf[255];

/* Daemonize me. */
static void daemonize(void) {
  int i;
  pid_t pid;

  /* Run as my child. */
  pid=fork();
  if (pid<0) exit(1); /* fork error */
  if (pid>0) exit(0); /* parent exits */

  /* Obtain a new process group. */
  (void)setsid();

  /* Close all open descriptors. */
  for(i=getdtablesize();i>=0;--i)
      (void)close(i);

  /* Reopen STDIN, STDOUT, STDERR to /dev/null. */
  i=open("/dev/null",O_RDWR);	/* STDIN */
  /*@ -sefparams @*/
  assert(dup(i) != -1); 	/* STDOUT */
  assert(dup(i) != -1);		/* STDERR */

  /* Know thy mask. */
  (void)umask(0x033);

  /* Run from a known spot. */
  assert(chdir("/") != -1);
  /*@ +sefparams @*/

  /* Catch child sig */
  (void)signal(SIGCHLD,SIG_IGN);

  /* Ignore tty signals */
  (void)signal(SIGTSTP,SIG_IGN);
  (void)signal(SIGTTOU,SIG_IGN);
  (void)signal(SIGTTIN,SIG_IGN);
}

/* open the serial port and set it up */
static void open_serial(char* device)
{
    /* 
     * Open modem device for reading and writing and not as controlling
     * tty.
     */
    if ((fd_out = open(device, O_RDWR|O_NOCTTY)) < 0) {
	fprintf(stderr, "gpspipe: error opening serial port\n");
	exit(1);
    }

    /* Save current serial port settings for later */
    if (tcgetattr(fd_out, &oldtio) != 0) {
	fprintf(stderr, "gpspipe: error reading serial port settings\n");
	exit(1);
    }

    /* Clear struct for new port settings. */
    /*@i@*/bzero(&newtio, sizeof(newtio));

    /* make it raw */
    (void)cfmakeraw(&newtio);
    /* set speed */
    /*@i@*/(void)cfsetospeed(&newtio, BAUDRATE);

    /* Clear the modem line and activate the settings for the port. */
    (void)tcflush(fd_out,TCIFLUSH);
    if (tcsetattr(fd_out,TCSANOW,&newtio) != 0) {
	(void)fprintf(stderr, "gpspipe: error configuring serial port\n");
	exit(1);
    }
}

static void usage(void)
{
    (void)fprintf(stderr, "Usage: gpspipe [OPTIONS] [server[:port[:device]]]\n\n"
		  "SVN ID: $Id: gpspipe.c 5053 2009-01-21 11:44:35Z esr $ \n"
		  "-d Run as a daemon.\n"
		  "-f [file] Write output to file.\n"
		  "-h Show this help.\n"
		  "-r Dump raw NMEA.\n"
		  "-R Dump super-raw mode (GPS binary).\n"
		  "-w Dump gpsd native data.\n"
		  "-j Turn on server-side buffering.\n"
		  "-l Sleep for ten seconds before connecting to gpsd.\n"
		  "-t Time stamp the data.\n"
		  "-s [serial dev] emulate a 4800bps NMEA GPS on serial port (use with '-r').\n"
		  "-n [count] exit after count packets.\n"
		  "-v Print a little spinner.\n"
		  "-V Print version and exit.\n\n"
		  "You must specify one, or both, of -r/-w.\n"
		  "You must use -f if you use -d.\n"
	);
}

int main( int argc, char **argv)
{
    int sock = 0;
    char buf[4096];
    ssize_t wrote = 0;
    bool timestamp = false;
    bool daemon = false;
    bool binary = false;
    bool sleepy = false;
    bool new_line = true;
    long count = -1;
    int option;
    unsigned int vflag = 0, l = 0;
    FILE * fp;

    char *arg = NULL, *colon1, *colon2, *device = NULL;
    char *port = DEFAULT_GPSD_PORT, *server = "127.0.0.1";
    char *serialport = NULL;
    char *filename = NULL;

    buf[0] = '\0';
    while ((option = getopt(argc, argv, "?dlhrRwjtvVn:s:f:")) != -1) {
	switch (option) {
	case 'n':
	    count = strtol(optarg, 0, 0);
	    break;
	case 'r':
	    (void)strlcat(buf, "r=1;", sizeof(buf));
	    break;
	case 'R':
	    binary=true;
	    (void)strlcat(buf, "r=2;", sizeof(buf));
	    break;
	case 'd':
	    daemon = true;
	    break;
	case 'l':
	    sleepy = true;
	    break;
	case 't':
	    timestamp = true;
	    break;
	case 'v':
	    vflag++;
	    break;
	case 'w':
	    (void)strlcat(buf, "w=1;", sizeof(buf));
	    break;
	case 'j':
	    (void)strlcat(buf, "j=1;", sizeof(buf));
	    break;
	case 'V':
	    (void)fprintf(stderr, "%s: SVN ID: $Id: gpspipe.c 5053 2009-01-21 11:44:35Z esr $ \n", argv[0]);
	    exit(0);
	case 's':
	    serialport = optarg;
	    break;
	case 'f':
	    filename = optarg;
	    break;
	case '?':
	case 'h':
	default:
	    usage();
	    exit(1);
	}
    }

    if (serialport!=NULL && strstr(buf, "r=1")==NULL) {
	(void)fprintf(stderr, "gpsipipe: use of '-s' requires '-r'.\n");
	exit(1);
    }

    if (filename==NULL && daemon) {
	(void)fprintf(stderr, "gpsipipe: use of '-d' requires '-f'.\n");
	exit(1);
    }

    if (strstr(buf, "r=")==NULL && strstr(buf, "w=1")==NULL) {
	(void)fprintf(stderr, "gpspipe: one of '-R', '-r' or '-w' is required.\n");
	exit(1);
    }
    /* Grok the server, port, and device. */
    /*@ -branchstate @*/
    if (optind < argc) {
	arg = strdup(argv[optind]);
	/*@i@*/colon1 = strchr(arg, ':');
	server = arg;
	if (colon1 != NULL) {
	    if (colon1 == arg) {
		server = NULL;
	    } else {
		*colon1 = '\0';
	    }
	    port = colon1 + 1;
	    colon2 = strchr(port, ':');
	    if (colon2 != NULL) {
		if (colon2 == port) {
		    port = NULL;
		} else {
		    *colon2 = '\0';
		}
		device = colon2 + 1;
		(void)snprintf(buf, sizeof(buf), "%sF=%s\n", buf, device);
	    }
	}
	colon1 = colon2 = NULL;
    }
    /*@ +branchstate @*/

    /* Daemonize if the user requested it. */
    if (daemon)
      daemonize();

    /* Sleep for ten seconds if the user requested it. */
    if (sleepy)
	(void)sleep(10);

    /* Open the output file if the user requested it.  If the user
       requested '-R', we use the 'b' flag in fopen() to "do the right
       thing" in non-linux/unix OSes. */
    if (filename==NULL) {
      fp = stdout;
    } else {
      if (binary)
	fp = fopen(filename,"wb");
      else
	fp = fopen(filename,"w");

      if (fp == NULL) {
	(void)fprintf(stderr,
		      "gpspipe: unable to open output file:  %s\n",
		      filename);
	exit(1);
      }
    }

    /* Open the serial port and set it up. */
    if (serialport)
	open_serial(serialport);

    /*@ -nullpass @*/
    sock = netlib_connectsock( server, port, "tcp");
    if (sock < 0) {
	(void)fprintf(stderr,
		      "gpspipe: could not connect to gpsd %s:%s, %s(%d)\n",
		      server, port, strerror(errno), errno);
	exit(1);
    }
    /*@ +nullpass @*/

    /* ship the assembled options */
    wrote = write(sock, buf, strlen(buf));
    if ((ssize_t)strlen(buf) != wrote) {
	(void)fprintf(stderr, "gpspipe: write error, %s(%d)\n",
		      strerror(errno), errno);
	exit(1);
    }

    if ((isatty(STDERR_FILENO) == 0) || daemon)
	vflag = 0;

    for(;;) {
	int i = 0;
	int j = 0;
	int readbytes = 0;

	if (vflag)
	    spinner(vflag, l++);
	readbytes = (int)read(sock, buf, sizeof(buf));
	if (readbytes > 0) {
	    for (i = 0 ; i < readbytes ; i++) {
		char c = buf[i];
		if (j < (int)(sizeof(serbuf) - 1)) {
		    serbuf[j++] = buf[i];
		}
		if (new_line && timestamp) {
		    time_t now = time(NULL);

		    new_line = 0;
		    if (fprintf(fp, "%.24s :", ctime(&now)) <= 0) {
			(void)fprintf(stderr,
				      "gpspipe: write error, %s(%d)\n",
				      strerror(errno), errno);
			exit(1);
		    }
		}
		if (fputc(c, fp) == EOF) {
		    fprintf( stderr, "gpspipe: Write Error, %s(%d)\n",
			     strerror(errno), errno);
		    exit(1);
		}

		if (c == '\n') {
		    if (serialport != NULL) {
			if (write(fd_out, serbuf, (size_t)j) == -1) {
			    fprintf(stderr,
				    "gpspipe: Serial port write Error, %s(%d)\n",
				     strerror(errno), errno);
			    exit(1);
			}
			j = 0;
		    }

		    new_line = true;
		    /* flush after every good line */
		    if (fflush(fp)) {
			(void)fprintf(stderr, "gpspipe: fflush Error, %s(%d)\n",
				strerror(errno), errno);
			exit(1);
		    }
		    if (count > 0) {
			if (0 >= --count) {
			    /* completed count */
			    exit(0);
			}
		    }
		}
	    }
	} else if (readbytes < 0) {
	    (void) fprintf(stderr, "gpspipe: read error %s(%d)\n",
			    strerror(errno), errno);
	    exit(1);
	}
    }

#ifdef __UNUSED__
    if (serialport != NULL) {
	/* Restore the old serial port settings. */
	if (tcsetattr(fd_out, TCSANOW, &oldtio) != 0) {
	    (void)fprintf(stderr, "Error restoring serial port settings\n");
	    exit(1);
	}
    }

    exit(0);
#endif /* __UNUSED__ */
}

static void spinner (unsigned int v, unsigned int num) {
    char *spin = "|/-\\";

    (void)fprintf(stderr, "\010%c", spin[(num/(1<<(v-1))) % 4]);
    (void)fflush(stderr);
    return;
}
